### DISCUSSION TITLE
8051 language files trivia

### DISCUSSION THREAD

**Discussion by gtackett** [NationalSecurityAgency/ghidra#5878] [General]
- **[Archimedes 8051-IDE](https://web.archive.org/web/20010620195905/http://www.archimedessoftware.com/8051.html)** was an early IDE (available prior to perhaps June 2001) with integrated **C51** compiler  that supported many of the then-existing chips derived from Intel's MCS-251 architecture for the original Intel 8051. According to my information, Archimedes was later acquired by Keil, now part of ARM Keil, apparently re-badging the Archimedes compiler.   Because the calling convention documented in Keil's documentation for its later Cx51 compilers differs significantly from that implemented in Ghidra for the "Archimedes" compiler, I suspect that after its acquisition of Archimedes, Keil later reengineered the compiler.
- **mx51** in Ghidra's language files refers to _some_ compiler (I'm guessing Keil's) targeting the NXP/Philips Semiconductor derivative of the 8051, featuring a 24-bit memory address space and faster instruction execution. I suspect this Ghidra language may not be entirely well suited for other 8051 variants having a smaller address space.

**Comment by PennRobotics** [comment]
For any curious minds: PSoC 3 (by Cypress, now Infineon, chips starting with CY8C3) might use the same memory layout as mx51.
